text/0000-crates-io-token-scopes.md

```diff
@@ -0,0 +1,223 @@
+the time of writing this RFC are `publish-new`, `publish-update`, `yank` and
+`change-owners`. The user will have to explicitly opt into extra scopes or the
+legacy permission model.
+
+Tokens created before the implementation of this RFC will use the legacy
+permission model.
+
+## Crates scope
+
+The user will be able to opt into limiting which crates the token can act on by
+defining a crates scope. It will be possible to set a crates scope even with
+the legacy endpoint scope.
+
+The crates scope can be left empty to allow the token to act on all the crates
+owned by the user, or it can contain the comma-separated list of crate names
+the token can interact with. Crate names can contain `*` to match one or more
+characters.
+
+For example, a crates scope of `serde,serde-*` allows the token to act on the
+`serde` crate or any present or future crates starting with `serde-`, if the
+user is an owner of those crates.
+
+The crates scope will allow access to all present and future crates matching
+it. When an endpoint that doesn't interact with crates is called by a token
+with a crates scope, the crates scope will be ignored and the call will be
+authorized.
+
+Tokens created before the implementation of this RFC won't have a crates scope,
+and it will be possible to use a crates scope in a token with the legacy
+endpoint scope.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Endpoint scopes and crates scope are two completly separate systems, and can be
+used independently from one another. Token scopes will be implemented entirely
+on the crates.io side, and there will be no change to `cargo` or alternate
+registries.
+
+## Endpoint scopes
+
+The scopes proposed by this RFC allow access to the following endpoints:
+
+| Endpoint | Required scope |
+| --- | --- |
+| `PUT /crates/new` (new crates) | **publish-new** |
+| `PUT /crates/new` (existing crates) | **publish-update** |
+| `DELETE /crates/:crate_id/:version/yank` | **yank** |
+| `PUT /crates/:crate_id/:version/unyank` | **yank** |
+| `PUT /crates/:crate_id/owners` | **change-owners** |
+| `DELETE /crates/:crate_id/owners` | **change-owners** |
+
+Removing an endpoint from a scope or adding an existing endpoint to an existing
+scope will be considered a breaking change. Adding newly created endpoints to
+an existing scope will be allowed only at the moment of their creation, if the
+crates.io team believes the new endpoint won't grant more privileges than the
+existing set of endpoints in that scope.
+
+## Crates scope
+
+The pattern for the crate scope is desugared into a regular expression,
+following these rules:
+
+* **`^`** is added at the start of the pattern, and **`$`** is added at the end of it.
+* **`,`** is desugared into `|`, separating multiple patterns.
+* **`*`** is desugared into `.+`, matching one or more characters greedily.
+* All other characters are quoted to prevent them from having a special meaning.
```

Comment:
An alternative would be to allow only characters that are allowed in crate names in addition to `,` and `*`. Since none of the characters in crate names have a special meaning in regular expressions, and `,` and `*` are not allowed in crate names, this would remove any ambiguities.

Reply:
True. I'm not sure changing this matters that much though.

Reply:
Yeah, this is more of an implementation commment – what I suggested is easier to translate into code, but it doesn't really matter.